internal/provider/custom_domain_resource.go

```diff
@@ -0,0 +1,267 @@
+	// Read the data from the state
+	var state customDomainResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the main env's custom domain info
+	environment := state.Environment.ValueString()
+
+	// isSwitching := state.IsPending.ValueBool()
+	customDomainInfo, err := r.client.GetCustomDomainInfo(environment, false)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error getting custom domain info",
+			"Could not get custom domain info, unexpected error: " + err.Error(),
+		)
+		return
+	}
+
+	isDomainOrSubdomainChanged := state.Domain.ValueString() != customDomainInfo.Domain || state.Subdomain.ValueStringPointer() != customDomainInfo.Subdomain
+
+	isSwitching := isDomainOrSubdomainChanged && customDomainInfo.IsVerified
+	if isSwitching {
+		// If the domain is switching, fetch the pending state instead.
+		customDomainInfo, err = r.client.GetCustomDomainInfo(environment, true)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Error getting custom domain info",
+				"Could not get custom domain info, unexpected error: " + err.Error(),
+			)
+			return
+		}
+	}
+
+	// Write logs using the tflog package
+	// Documentation: https://terraform.io/plugin/log
+	tflog.Trace(ctx, "read a custom_domain resource")
+
+	// Set the data from the state into the response
+	state.Domain = types.StringValue(customDomainInfo.Domain)
+	state.Subdomain = types.StringPointerValue(customDomainInfo.Subdomain)
+
+	// So as not to need an update after verification of a domain, 
+	// these fields are only updated if the domain is not verified, which is when they are
+	// returned.
+	if customDomainInfo.TxtRecordKey != nil {
+		state.TxtRecordKey = types.StringPointerValue(customDomainInfo.TxtRecordKey)
+	}
+	if customDomainInfo.TxtRecordValue != nil {
+		state.TxtRecordValue = types.StringPointerValue(customDomainInfo.TxtRecordValue)
+	}
+	if customDomainInfo.CnameRecordKey != nil {
+		state.CnameRecordKey = types.StringPointerValue(customDomainInfo.CnameRecordKey)
+	}
+	if customDomainInfo.CnameRecordValue != nil {
+		state.CnameRecordValue = types.StringPointerValue(customDomainInfo.CnameRecordValue)
+	}
+	
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (r *customDomainResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Read Terraform plan data into the model
+	var plan customDomainResourceModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Read the current state data
+	var state customDomainResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Re-fetch the main env's custom domain info to check 
+	// if its verification status has changed.
+	customDomainInfo, err := r.client.GetCustomDomainInfo(state.Environment.ValueString(), false)
```

Comment:
since the `environment` attribute requires replace, we can just use the plan for this